### PR TITLE
[IOTDB-4437] Implement event window in RawDataAggregationOperator

### PR DESCRIPTION
**Pull request by ZhanGHanG9991**
## Description
Support event window aggregation on RawDataAggregationOperator

<hr>

This PR has:
- [x] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [x] added or updated version, __license__, or notice information
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowManagerFactory.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+package org.apache.iotdb.db.mpp.execution.operator.window;
+
+import org.apache.iotdb.db.mpp.aggregation.timerangeiterator.ITimeRangeIterator;
+import org.apache.iotdb.db.mpp.execution.operator.exception.UnSupportedEventAggregationMeasureTypeException;
+import org.apache.iotdb.db.mpp.execution.operator.exception.UnSupportedEventAggregationWindowTypeException;
+import org.apache.iotdb.tsfile.exception.write.UnSupportedDataTypeException;
+
+public class WindowManagerFactory {
+
+  public IWindowManager genWindowManager(
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowManagerFactory.java:29]:
```suggestion
  public static IWindowManager genWindowManager(
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.execution.operator.window;
+
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+
+public class WindowParameter {
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java:24]: The WindowParameter class can be used as an abstract class, and different types of window parameters extend this class to reduce unnecessary parameters.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventWindowManager.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    return true;
+  }
+
+  @Override
+  public boolean isTsBlockOutOfBound(TsBlock inputTsBlock) {
+    return false;
+  }
+
+  @Override
+  public TsBlockBuilder createResultTsBlockBuilder(List<Aggregator> aggregators) {
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventWindowManager.java:96]: It seems that the case where we need output start time and end time are not considered here.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventWindow.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.apache.iotdb.db.mpp.aggregation.Accumulator;
+import org.apache.iotdb.tsfile.read.common.block.TsBlock;
+import org.apache.iotdb.tsfile.read.common.block.column.Column;
+
+public abstract class EventWindow implements IWindow {
+
+  protected WindowParameter windowParameter;
+
+  protected long startTime;
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventWindow.java:30]: How do we record and output the end time?

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventLongWindowManager.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      TsBlockBuilder resultTsBlockBuilder, List<Aggregator> aggregators) {
+    // If we have appended aggregation result to resultTsBlockBuilder, we need return directly.
+    if (this.hasAppendedResult) {
+      return;
+    }
+    // Append aggregation results to valueColumnBuilders.
+    ColumnBuilder[] columnBuilders =
+        appendOriginAggregationResult(resultTsBlockBuilder, aggregators);
+    // Judge whether we need output event column.
+    if (windowParameter.isNeedOutputEvent()) {
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventLongWindowManager.java:45]: It seems that the case where we need output start time and end time are not considered here.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/exception/UnSupportedEventAggregationMeasureTypeException.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.execution.operator.exception;
+
+import org.apache.iotdb.db.mpp.execution.operator.window.CompareType;
+
+public class UnSupportedEventAggregationMeasureTypeException extends RuntimeException {
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/exception/UnSupportedEventAggregationMeasureTypeException.java:24]: You can use `IllegalArgumentException` directly, we will check for correct semantics in the Analyzer phase.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/exception/UnSupportedEventAggregationWindowTypeException.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.execution.operator.exception;
+
+import org.apache.iotdb.db.mpp.execution.operator.window.WindowType;
+
+public class UnSupportedEventAggregationWindowTypeException extends RuntimeException {
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/exception/UnSupportedEventAggregationWindowTypeException.java:24]: You can use `IllegalArgumentException` directly, we will check for correct semantics in the Analyzer phase.

**Review by liuminghui233 — CHANGES_REQUESTED** [review]

**Review by ycycse — COMMENTED** [review]
can the window and windowManager of EqualEventDouble,EqualEventLong,EqualEventFloat and EqualEventInt be replaced with the variationEvent ones with the delta value of zero?

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventBooleanWindowManager.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  public EventBooleanWindowManager(WindowParameter windowParameter, boolean ascending) {
+    super(windowParameter, ascending);
+  }
+
+  @Override
+  public void appendAggregationResult(
+      TsBlockBuilder resultTsBlockBuilder, List<Aggregator> aggregators) {
+    // If we have appended aggregation result to resultTsBlockBuilder, we need return directly.
+    if (this.hasAppendedResult) {
```
Comment (ycycse) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventBooleanWindowManager.java:38]: why hasAppendedResult?

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/VariationEventDoubleWindowManager.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    int i = 0, size = inputTsBlock.getPositionCount();
+    for (; i < size; i++) {
+      if (!controlColumn.isNull(i)
+          && Math.abs(
+                  controlColumn.getDouble(i)
+                      - ((VariationEventDoubleWindow) eventWindow).getPreviousEventValue())
+              > windowParameter.getDelta()) {
+        break;
+      }
+    }
```
Comment (ycycse) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/VariationEventDoubleWindowManager.java:52]:
```suggestion
    int i = 0, size = inputTsBlock.getPositionCount();
    double previousEventValue = ((VariationEventDoubleWindow) eventWindow).getPreviousEventValue();
    for (; i < size; i++) {
      if (!controlColumn.isNull(i) && Math.abs(controlColumn.getDouble(i) - previousEventValue > windowParameter.getDelta()) {
        break;
      }
    }
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowManagerFactory.java (hunk 1 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowManagerFactory.java:29]: I've modified it.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java (hunk 2 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java:24]: I've made WindowParameter an abstract class, and added TimeWindowParameter and EventWindowParameter.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventWindow.java (hunk 4 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventWindow.java:30]: We've resolved this issue.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventWindowManager.java (hunk 3 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventWindowManager.java:96]: I've added it.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventLongWindowManager.java (hunk 5 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventLongWindowManager.java:45]: I've added them.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/exception/UnSupportedEventAggregationWindowTypeException.java (hunk 7 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/exception/UnSupportedEventAggregationWindowTypeException.java:24]: I've changed them to IllegalArgumentException.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/exception/UnSupportedEventAggregationMeasureTypeException.java (hunk 6 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/exception/UnSupportedEventAggregationMeasureTypeException.java:24]: I've changed them to IllegalArgumentException.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EqualEventBooleanWindow.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.execution.operator.window;
+
+import org.apache.iotdb.tsfile.read.common.block.column.Column;
+
+public class EqualEventBooleanWindow extends EventBooleanWindow {
```
[the remark this answers is not in the document]
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EqualEventBooleanWindow.java:24]: OK. I will use freemarker to generate code in the future.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventBooleanWindowManager.java (hunk 8 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventBooleanWindowManager.java:38]: I've deleted this.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/VariationEventDoubleWindowManager.java (hunk 9 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/VariationEventDoubleWindowManager.java]: I've modified them.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/TimeWindowManager.java [hunk 11]
```diff
@@ -35,10 +42,18 @@ public class TimeWindowManager implements IWindowManager {
 
   private boolean needSkip;
 
-  public TimeWindowManager(ITimeRangeIterator timeRangeIterator) {
+  private TimeWindowParameter timeWindowParameter;
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/TimeWindowManager.java:45]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/RawDataAggregationOperator.java [hunk 12]
```diff
@@ -46,19 +46,23 @@ public class RawDataAggregationOperator extends SingleInputAggregationOperator {
 
   private final IWindowManager windowManager;
 
+  private boolean isSkipping = false;
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/RawDataAggregationOperator.java:49]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    this.controlColumnIndex = controlColumnIndex;
+  }
+
+  public boolean isNeedOutputEndTime() {
+    return needOutputEndTime;
+  }
+
+  public void setNeedOutputEndTime(boolean needOutputEndTime) {
+    this.needOutputEndTime = needOutputEndTime;
+  }
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java:66]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    this.dataType = dataType;
+  }
+
+  public int getControlColumnIndex() {
+    return controlColumnIndex;
+  }
+
+  public void setControlColumnIndex(int controlColumnIndex) {
+    this.controlColumnIndex = controlColumnIndex;
+  }
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java:58]: done,

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    return windowType;
+  }
+
+  public TSDataType getDataType() {
+    return dataType;
+  }
+
+  public void setDataType(TSDataType dataType) {
+    this.dataType = dataType;
+  }
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java:50]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+public abstract class WindowParameter {
+
+  protected WindowType windowType;
+
+  private TSDataType dataType;
+
+  private int controlColumnIndex;
+
+  private boolean needOutputEndTime;
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java:32]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+
+public abstract class WindowParameter {
+
+  protected WindowType windowType;
+
+  private TSDataType dataType;
+
+  private int controlColumnIndex;
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/WindowParameter.java:30]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/OperatorTreeGenerator.java [hunk 18]
```diff
@@ -1391,13 +1393,16 @@ public Operator visitAggregation(AggregationNode node, LocalExecutionPlanContext
           calculateMaxAggregationResultSize(
               aggregationDescriptors, timeRangeIterator, context.getTypeProvider());
 
+      WindowParameter windowParameter = new TimeWindowParameter(TSDataType.INT64, 0, false);
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/OperatorTreeGenerator.java:1396]: done

File: server/src/test/java/org/apache/iotdb/db/mpp/execution/operator/OperatorMemoryTest.java [hunk 19]
```diff
@@ -1405,14 +1407,17 @@ public void rawDataAggregationOperatorTest() throws IllegalPathException {
         AggregationUtil.calculateMaxAggregationResultSize(
             aggregationDescriptors, timeRangeIterator, typeProvider);
 
+    WindowParameter windowParameter = new TimeWindowParameter(TSDataType.INT64, 0, false);
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/test/java/org/apache/iotdb/db/mpp/execution/operator/OperatorMemoryTest.java:1410]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventBooleanWindow.java [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  }
+
+  public boolean getEventValue() {
+    return eventValue;
+  }
+
+  public void setEventValue(boolean eventValue) {
+    this.eventValue = eventValue;
+  }
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventBooleanWindow.java:65]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventBooleanWindow.java [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+  public boolean getPreviousEventValue() {
+    return previousEventValue;
+  }
+
+  public void setPreviousEventValue(boolean previousEventValue) {
+    this.previousEventValue = previousEventValue;
+  }
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/EventBooleanWindow.java:73]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/SlidingWindowAggregationOperator.java [hunk 22]
```diff
@@ -24,8 +24,12 @@
 import org.apache.iotdb.db.mpp.execution.operator.Operator;
 import org.apache.iotdb.db.mpp.execution.operator.OperatorContext;
 import org.apache.iotdb.db.mpp.plan.planner.plan.parameter.GroupByTimeParameter;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
 import org.apache.iotdb.tsfile.read.common.TimeRange;
+import org.apache.iotdb.tsfile.read.common.block.TsBlockBuilder;
 
+import java.util.ArrayList;
+import java.util.Arrays;
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/SlidingWindowAggregationOperator.java:32]: SlidingWindow share a different framework with RawDataAggregator.
So I think realize it in another PR is better.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/IWindowManager.java [hunk 23]
```diff
@@ -88,4 +88,27 @@ public interface IWindowManager {
...
+      dataTypes.addAll(Arrays.asList(aggregator.getOutputType()));
+    }
+    return dataTypes;
+  }
+
+  TsBlockBuilder createResultTsBlockBuilder(List<Aggregator> aggregators);
+
+  void appendAggregationResult(TsBlockBuilder resultTsBlockBuilder, List<Aggregator> aggregators);
+
+  boolean notInitedLastTimeWindow();
```
[the remark this answers is not in the document]
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/window/IWindowManager.java]: Done.

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**